### DISCUSSION TITLE
Ibis for Pandas users

### DISCUSSION THREAD

**Discussion by kesmit13** [ibis-project/ibis#4647] [Proposals]
You have a section in the documentation for SQL users coming to Ibis, but I think it would be helpful to have one for Pandas users as well. It took me a while to figure out how to do some simpler things in Ibis like using `mutate` to add a column, the fact that there is no `shape` or `dtypes` attributes, just the general idea that there is no way to do an in-place operation (everything returns a new expression). Most are little things, but they aren't always obvious to new users.

**Comment by gforsyth** [comment]
I think this is a great idea @kesmit13 !  If you have any other things on your list of pain points / things that you expected and can document them here that would be a huge help for us.

**Comment by kesmit13** [comment]
I can make a list of things. I do think it would be simpler to just add things like `dtypes`, `shape`, `unique`, etc. rather than document why they aren't there. The implementations are quite simple. Is there a philosophical reason for not adding those types of things?

**Reply by gforsyth** [reply]
Not necessarily, but one possible philosophical barrier here is that we don't want to have multiple methods/attributes that do the same thing but have a different name.  

Some, like `shape`, might break Ibis' contract a bit.  I imagine that it would either execute eagerly, which most things in Ibis do not, or it would return something that doesn't really fit into a chained expression.

We're interested in making things comfortable, and in adding in missing functionality, but not adding in methods just to match pandas.

**Reply by cpcloud** [reply]
The list would be useful to have so that we don't add a lot of functionality that isn't used. Some functionality just doesn't fit into the operating model of ibis, like unique on a column. That's because it doesn't compose: e.g., what would `t.a.distinct() + t.b` return?

**Reply by kesmit13** [reply]
I guess I feel like `t.a.distinct()` would return the value of `t.a.to_projection().distinct()[0]`. Similarly, `t.a.head()` would return `t.a.to_projection().head()[0]`. To me, it seems odd to have those operations only on the table when they are very commonly used on columns as well. Obviously, I can type the substitution that I did above, but it's non-obvious code.

**Reply by kesmit13** [reply]
I was working with some notebooks today and realized that the following code is effectively the same as calling `distinct` on a column. I'd still like to see `distinct` implemented on the column itself...
```
t.a.value_counts().a
```

**Reply by cpcloud** [reply]
@kesmit13 I understand the desire to have `.distinct()` on a column. For reasons I talk about in #1711, mainly that `t.col.distinct()` doesn't compose in general, I removed that method in #3545.

**Reply by jcmkk3** [reply]
> t.a.value_counts().a

I'm curious what are you doing with the results once you perform this operation? Are you feeding them into some other function? Maybe that would help understand your use case better.

Another alternative to what you've written is to select a single column table, then do a distinct on that.
```python
t[["a"]].distinct()
# or
t.select("a").distinct()
```

**Reply by kesmit13** [reply]
It's not necessarily used to perform more operations on (except maybe count), but I sometimes just need to get the distinct values of a column to populate a form, or to do an `in` operation on the client. The code above isn't quite the same as `distinct` on a column since calling `execute` returns a `DataFrame` rather than a `Series`. Although, you could add a `.a` to the end to get that.

**Reply by jcmkk3** [reply]
You are right. They do not return the same objects. That's why I was wondering how you were using it after. I see how it could be convenient in some contexts, but I agree with @cpcloud that it doesn't compose very in a `ColumnExpr`. I feel the same way about `value_counts` and `topk`. `value_counts` at least returns a `Table` instead of the weird `TopK` object.

**Reply by cpcloud** [reply]
We are planning to address the topk object problem, but not entirely sure what the best way forward is yet. Topk can be used alone and can also be used as a filter, which effectively IIRC adds a semi-join to the "topk-alone" variety, to show the rows from the original table that match the top K keys.

**Reply by cpcloud** [reply]
Let's discuss that elsewhere though to avoid hijacking this thread :)

**Reply by kesmit13** [reply]
I'd be interested in that topic as well. I wrote a version of pandas' `describe` method that uses existing Ibis operations including `topk`, but it did complicate things.

**Comment by cpcloud** [comment]
It seems like we can resolve this discussion since @kesmit13 contributed a wonderful [Ibis for Pandas Users notebook](https://ibis-project.org/docs/dev/ibis-for-pandas-users/)!
